### PR TITLE
Backport #378: PipeBlob as command's in/out

### PR DESCRIPTION
**Pull request by Ingvord**
This PR backports #378 into tango-9-lts branch. Client may now use PipeBlob as command argument and server may return PipeBlob as command result.

### CODE REVIEW

**Review by Ingvord — COMMENTED** [review]
SonarQube analysis found issues:
Bug Bugs: 0
Vulnerability Vulnerabilities: 0
Code Smell Code Smells: 4

[See all issues in SonarCloud](https://sonarcloud.io/project/issues?id=org.tango-controls%3Acpp-tango%3Atango-9-lts&pullRequest=477&resolved=false)

File: cppapi/client/devapi_base.cpp [hunk 1]
```diff
@@ -1345,6 +1346,7 @@ DeviceData Connection::command_inout(string &command, DeviceData &data_in)
 // Now, try to execute the command
 //
 
+                //TODO extract class hierarchy
```
Comment (Ingvord) [inline cppapi/client/devapi_base.cpp]: Code Smell Code Smell: Complete the task associated to this "TODO" comment. ([cpp:S1135](https://sonarcloud.io/organizations/tango-controls/rules#rule_key=cpp%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=org.tango-controls%3Acpp-tango%3Atango-9-lts&pullRequest=477&issues=AWWPm0Fu3Qu1svVwsabD&open=AWWPm0Fu3Qu1svVwsabD)

[ISSUE_KEY]: AWWPm0Fu3Qu1svVwsabD

File: cppapi/client/devapi_base.cpp [hunk 2]
```diff
@@ -1212,6 +1212,7 @@ void Connection::set_timeout_millis(int millisecs)
 }
 
 
+//TODO all command_inout methods must be moved to DeviceProxy
```
Comment (Ingvord) [inline cppapi/client/devapi_base.cpp]: Code Smell Code Smell: Complete the task associated to this "TODO" comment. ([cpp:S1135](https://sonarcloud.io/organizations/tango-controls/rules#rule_key=cpp%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=org.tango-controls%3Acpp-tango%3Atango-9-lts&pullRequest=477&issues=AWWPm0Fu3Qu1svVwsabC&open=AWWPm0Fu3Qu1svVwsabC)

[ISSUE_KEY]: AWWPm0Fu3Qu1svVwsabC

File: cppapi/client/devapi_base.cpp [hunk 3]
```diff
@@ -5512,7 +5492,31 @@ void DeviceProxy::write_pipe(DevicePipe &dev_pipe)
...
+        result.data_blob.name = bl_name.c_str();
+
+
+    DevVarPipeDataEltArray *tmp_ptr = dev_pipe.get_root_blob().get_insert_data();
+    if (tmp_ptr == Tango_nullptr)
+    {
+        Except::throw_exception(API_PipeNoDataElement, "No data in pipe!", "DeviceProxy::write_pipe()");
+    }
+
+    DevULong max, len;
```
Comment (Ingvord) [inline cppapi/client/devapi_base.cpp]: Code Smell Code Smell: Define each identifier in a dedicated statement. ([cpp:SingleDeclarationPerStatement](https://sonarcloud.io/organizations/tango-controls/rules#rule_key=cpp%3ASingleDeclarationPerStatement))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=org.tango-controls%3Acpp-tango%3Atango-9-lts&pullRequest=477&issues=AWWPm0Fu3Qu1svVwsabB&open=AWWPm0Fu3Qu1svVwsabB)

[ISSUE_KEY]: AWWPm0Fu3Qu1svVwsabB

File: cppapi/client/devapi_base.cpp [hunk 4]
```diff
@@ -5544,6 +5548,7 @@ DevicePipe DeviceProxy::write_read_pipe(DevicePipe &pipe_data)
 // Prepare data sent to device
 //
 
+    //TODO replace with createDevPipeData
```
Comment (Ingvord) [inline cppapi/client/devapi_base.cpp]: Code Smell Code Smell: Complete the task associated to this "TODO" comment. ([cpp:S1135](https://sonarcloud.io/organizations/tango-controls/rules#rule_key=cpp%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=org.tango-controls%3Acpp-tango%3Atango-9-lts&pullRequest=477&issues=AWWPm0Fu3Qu1svVwsabE&open=AWWPm0Fu3Qu1svVwsabE)

[ISSUE_KEY]: AWWPm0Fu3Qu1svVwsabE

**Comment by bourtemb** [comment]
As it is implemented it is not convenient to use the DevicePipeBlob.
We would need to find a more user-friendly way to create and extract DevicePipeBlob.
Is this feature request still needed? It seems some users are using json strings as alternative...

**Pull request closed without merging**